File: app/src/main/java/com/arkapp/partyplanner/ui/guestList/GuestListFragment.java
package com.arkapp.partyplanner.ui.guestList;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;

import androidx.activity.OnBackPressedCallback;
import androidx.fragment.app.Fragment;
import androidx.navigation.ActivityKt;
import androidx.recyclerview.widget.RecyclerView;

import com.arkapp.partyplanner.R;
import com.arkapp.partyplanner.R.id;
import com.arkapp.partyplanner.data.models.PartyDetails;
import com.arkapp.partyplanner.data.repository.PrefRepository;
import com.arkapp.partyplanner.data.room.AppDatabase.Companion;
import com.arkapp.partyplanner.data.room.HistorySummaryDao;
import com.arkapp.partyplanner.data.room.SummaryDetailsDao;
import com.arkapp.partyplanner.utils.AppDataKt;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.arkapp.partyplanner.utils.AppDataKt.OPTION_CHECKLIST;
import static com.arkapp.partyplanner.utils.AppDataKt.OPTION_CREATE;
import static com.arkapp.partyplanner.utils.ViewUtilsKt.initVerticalAdapter;
import static com.arkapp.partyplanner.utils.ViewUtilsKt.toast;
import static java.sql.DriverManager.println;

public final class GuestListFragment extends Fragment {
    private PrefRepository prefRepository;

    public GuestListFragment() {
        super(R.layout.fragment_guest_list);
    }

    public void onViewCreated(@NotNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        prefRepository = new PrefRepository(requireContext());

        AppDataKt.setOPENED_GUEST_LIST(true);

        GuestListAdapter adapter = new GuestListAdapter(AppDataKt.getGUEST_LIST_NAMES());
        RecyclerView rv = getView().findViewById(id.guestListRv);
        initVerticalAdapter(rv, adapter, true);

        //Handling back click
        OnBackPressedCallback callback = new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                // Handle the back button event
                println("guest list back pressed");
                PartyDetails partyDetails = prefRepository.getCurrentPartyDetails();
                partyDetails.setGuestNameList(AppDataKt.getGUEST_LIST_NAMES());
                prefRepository.setCurrentPartyDetails(partyDetails);

                //Storing the guest check list in the SQL when back is pressed.
                this.remove();
                if (AppDataKt.getCURRENT_SELECTED_OPTION() == OPTION_CHECKLIST || AppDataKt.getCURRENT_SELECTED_OPTION() == OPTION_CREATE) {
                    toast(requireContext(), "Please wait... Saving data!");
                    new UpdateSummaryAsyncTask(requireActivity(), prefRepository).execute();
                } else
                    new UpdateHistorySummaryAsyncTask(requireActivity(), prefRepository).execute();
            }
        };
        requireActivity().getOnBackPressedDispatcher().addCallback(getViewLifecycleOwner(), callback);
    }

    private final class UpdateSummaryAsyncTask extends AsyncTask<Void, Void, Void> {
        private final Activity context;
        private final PrefRepository prefRepository;

        public UpdateSummaryAsyncTask(@NotNull Activity context, @NotNull PrefRepository prefRepository) {
            super();
            this.context = context;
            this.prefRepository = prefRepository;
        }

        @Nullable
        @Override
        protected Void doInBackground(@NotNull Void... params) {
            SummaryDetailsDao summaryDao = (new Companion()).getDatabase(context).summaryDao();

            summaryDao.delete(prefRepository.getCurrentUser().getUid());
            summaryDao.insert(AppDataKt.convertSummary(prefRepository.getCurrentPartyDetails(), prefRepository.getCurrentUser().getUid()));
            return null;
        }

        @SuppressLint({"SetTextI18n"})
        @Override
        protected void onPostExecute(@Nullable Void summaryData) {
            ActivityKt.findNavController(this.context, R.id.fragment).navigate(R.id.action_guestListFragment_to_finalChecklistFragment);
        }

    }

    private class UpdateHistorySummaryAsyncTask extends AsyncTask<Void, Void, Void> {
        private final Activity context;
        private final PrefRepository prefRepository;

        public UpdateHistorySummaryAsyncTask(@NotNull Activity context, @NotNull PrefRepository prefRepository) {
            super();
            this.context = context;
            this.prefRepository = prefRepository;
        }

        @Nullable
        @Override
        protected Void doInBackground(@NotNull Void... params) {
            HistorySummaryDao summaryDao = (new Companion()).getDatabase(context).historySummaryDao();

            summaryDao.delete(prefRepository.getCurrentPartyDetails().getId());
            summaryDao.insert(AppDataKt.convertHistorySummary(prefRepository.getCurrentPartyDetails(), prefRepository.getCurrentUser().getUid()));
            return null;
        }

        @SuppressLint({"SetTextI18n"})
        protected void onPostExecute(@Nullable Void summaryList) {
            ActivityKt.findNavController(this.context, id.fragment).navigate(R.id.action_guestListFragment_to_finalChecklistFragment);
        }
    }
}

File: app/src/main/java/com/arkapp/partyplanner/ui/guestList/GuestListFragment.kt
package com.arkapp.partyplanner.ui.guestList

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import android.os.Bundle
import android.view.View
import androidx.activity.addCallback
import androidx.fragment.app.Fragment
import androidx.navigation.findNavController
import com.arkapp.partyplanner.R
import com.arkapp.partyplanner.data.repository.PrefRepository
import com.arkapp.partyplanner.data.room.AppDatabase
import com.arkapp.partyplanner.utils.*
import kotlinx.android.synthetic.main.fragment_guest_list.*


class GuestListFragment : Fragment(R.layout.fragment_guest_list) {

    private val prefRepository by lazy { PrefRepository(requireContext()) }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        OPENED_GUEST_LIST = true

        val adapter = GuestListAdapter(GUEST_LIST_NAMES)
        guestListRv.initVerticalAdapter(adapter, true)

        //Handling the back click
        requireActivity()
            .onBackPressedDispatcher
            .addCallback(viewLifecycleOwner) {
                println("guest list back pressed")
                val partyDetails = prefRepository.currentPartyDetails
                partyDetails.guestNameList = GUEST_LIST_NAMES
                prefRepository.currentPartyDetails = partyDetails

                //Storing the guest check list in the SQL when back is pressed.
                this.remove()
                if (CURRENT_SELECTED_OPTION == OPTION_CHECKLIST || CURRENT_SELECTED_OPTION == OPTION_CREATE) {
                    requireContext().toast("Please wait... Saving data!")
                    UpdateHistorySummaryAsyncTask(requireActivity(), prefRepository).execute()
                } else
                    UpdateHistorySummaryAsyncTask(requireActivity(), prefRepository).execute()

                true
            }
    }

    //Storing summary in SQL
    private class UpdateSummaryAsyncTask(private val context: Activity,
                                         private val prefRepository: PrefRepository) : AsyncTask<Void, Void, Void?>() {

        override fun doInBackground(vararg params: Void?): Void? {
            val summaryDao = AppDatabase.Companion().getDatabase(context).summaryDao()
            summaryDao.delete(prefRepository.currentUser?.uid!!)
            summaryDao.insert(convertSummary(prefRepository.currentPartyDetails,
                                             prefRepository.currentUser?.uid!!))
            return null
        }

        @SuppressLint("SetTextI18n")
        override fun onPostExecute(summaryData: Void?) {
            context.findNavController(R.id.fragment)
                .navigate(R.id.action_guestListFragment_to_finalChecklistFragment)
        }
    }

    //Storing History in SQL
    private class UpdateHistorySummaryAsyncTask(private val context: Activity,
                                                private val prefRepository: PrefRepository) : AsyncTask<Void, Void, Void?>() {

        override fun doInBackground(vararg params: Void?): Void? {
            val summaryDao = AppDatabase.Companion().getDatabase(context).historySummaryDao()
            summaryDao.delete(prefRepository.currentPartyDetails.id!!)
            summaryDao.insert(convertHistorySummary(prefRepository.currentPartyDetails,
                                                    prefRepository.currentUser?.uid!!))
            return null
        }

        @SuppressLint("SetTextI18n")
        override fun onPostExecute(summaryList: Void?) {
            context.findNavController(R.id.fragment)
                .navigate(R.id.action_guestListFragment_to_finalChecklistFragment)
        }
    }

}
